### PR TITLE
[make:user] Legacy <= 5.3 & Doctrine Cleanup

### LINKED ISSUE #990: make:user & password_hashers: skip password_hashers update if possible
Hi!

The latest `security-bundle` recipe comes with a security.yaml file containing this out of the box:

```yaml
security:
    # ...

    # https://symfony.com/doc/current/security.html#c-hashing-passwords
    password_hashers:
        Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface: 'auto'
```

And so, when `make:user` is executed (with "yes" for password), we do not need to update the password hasher. In other words, what we do now is this:

```diff
security:
    password_hashers:
        Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface: 'auto'
+        App\Entity\User:
+            algorithm: auto
```

Those 2 new lines should not be added if `PasswordAuthenticatedUserInterface` is already present, as they are just not needed.

### PR DESCRIPTION
**Pull request by jrushlow**
- `TemplateComponentGenerator::generateUseStatements()` can now handle use statements with aliases.
- Add `doctrine/orm` conflict w/ `<2.10`
- drop legacy Symfony < `5.4` Doctrine && User conditionals
- remove legacy `UserBuilder` tests
- fixes #990 

- [x] Depends on #1104

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     "require-dev": {
         "composer/semver": "^3.0",
         "doctrine/doctrine-bundle": "^2.4",
-        "doctrine/orm": "^2.3",
+        "doctrine/orm": "^2.10.0",
```
Comment (jrushlow) [inline composer.json:31]: potentially controversial.. need to double check dependency dependencies to ensure this is possible... Reasoning being the bump to 10 / conflict w/ <2.9 is primarily related to the new exception classes introduced in 2.10 that will become interfaces in 3.0..

File: src/Security/SecurityConfigUpdater.php [hunk 2]
```diff
@@ -185,34 +183,33 @@ private function updateProviders(UserClassConfiguration $userConfig, string $use
         $this->manipulator->setData($newData);
     }
 
-    private function updatePasswordHashers(UserClassConfiguration $userConfig, string $userClass, string $keyName = 'password_hashers'): void
+    private function updatePasswordHashers(UserClassConfiguration $userConfig, string $userClass): void
     {
         $newData = $this->manipulator->getData();
-        if ('password_hashers' === $keyName && isset($newData['security']['encoders'])) {
-            // fallback to "encoders" if the user already defined encoder config
-            $this->updatePasswordHashers($userConfig, $userClass, 'encoders');
```
Comment (jrushlow) [inline src/Security/SecurityConfigUpdater.php:189]: Re for situations where a user still has `security.encoders` set - we could do any of:
A) Throw an exception  e.g. "MakerBundle does not support password encoders.."
B) Remove the existing `security.encoders` and continue with setting `security.password_hashers`

The problem with the latter is we dont have a way to tell the user that we're removing the existing encoders.. I suppose we _could_ start linking `make:user`'s IO to `SecurityConfigUpdater` or something to make this "async" but that just feels like a headache w/ not alot of gain. 

If we just throw an exception, what action do we tell the user to take to correct the problem?

File: composer.json (hunk 1 above)
Comment (weaverryan) [reply composer.json:31]: My initial thought is: it's ok. ORM 2.10 was released Oct 3rd, 2021. Maker now requires Symfony 5.4, which was released at the end of November 2021. So, if you have 5.4, you've been upgrading already. And bumping to 2.10 is a minor release bump.

File: src/DependencyInjection/CompilerPass/SetDoctrineManagerRegistryClassPass.php [hunk 3]
```diff
@@ -22,9 +22,9 @@ class SetDoctrineManagerRegistryClassPass implements CompilerPassInterface
...
     public function process(ContainerBuilder $container): void
     {
-        if ($container->hasAlias(ManagerRegistry::class)) {
-            $definition = $container->getDefinition('maker.entity_class_generator');
-            $definition->addMethodCall('setMangerRegistryClassName', [ManagerRegistry::class]);
-        }
+//        if ($container->hasAlias(ManagerRegistry::class)) {
+//            $definition = $container->getDefinition('maker.entity_class_generator');
+//            $definition->addMethodCall('setMangerRegistryClassName', [ManagerRegistry::class]);
+//        }
```
Comment (weaverryan) [inline src/DependencyInjection/CompilerPass/SetDoctrineManagerRegistryClassPass.php:28]: This isn't needed anymore? Can we delete?

File: src/Maker/MakeUser.php [hunk 4]
```diff
@@ -169,13 +169,22 @@ public function generate(InputInterface $input, ConsoleStyle $io, Generator $gen
         // C) Generate a custom user provider, if necessary
         if (!$userClassConfiguration->isEntity()) {
             $userClassConfiguration->setUserProviderClass($generator->getRootNamespace().'\\Security\\UserProvider');
+
+            $useStatements = [
+                UnsupportedUserException::class,
+                UsernameNotFoundException::class,
```
Comment (weaverryan) [inline src/Maker/MakeUser.php:175]: I think this is backwards: `UsernameNotFoundException` is deprecated and `UserNotFoundException` is the new one.

File: src/Security/SecurityConfigUpdater.php (hunk 2 above)
Comment (weaverryan) [reply src/Security/SecurityConfigUpdater.php:169]: If `encoders` is set, let's throw a clear exception that explains to change to `password_hashers`.

**Review by weaverryan — COMMENTED** [review]
Very complex, nice work! Minor comments

File: src/Security/SecurityConfigUpdater.php [hunk 5]
```diff
@@ -165,34 +163,33 @@ private function updateProviders(UserClassConfiguration $userConfig, string $use
...
 
+        // The security-bundle recipe sets the password hasher via Flex. If it exists, move on...
+        if (isset($newData['security']['password_hashers'][PasswordAuthenticatedUserInterface::class])) {
```
Comment (jrushlow) [inline src/Security/SecurityConfigUpdater.php:171]: merged in via #1110

File: src/DependencyInjection/CompilerPass/SetDoctrineManagerRegistryClassPass.php (hunk 3 above)
Comment (jrushlow) [reply src/DependencyInjection/CompilerPass/SetDoctrineManagerRegistryClassPass.php]: yup yup

File: src/Maker/MakeUser.php (hunk 4 above)
Comment (jrushlow) [reply src/Maker/MakeUser.php]: good catch

File: src/Security/SecurityConfigUpdater.php (hunk 2 above)
Comment (jrushlow) [reply src/Security/SecurityConfigUpdater.php:169]: sounds like a plan

**Review by weaverryan — APPROVED** [review]

**Comment by weaverryan** [comment]
HUGE work - thanks Jesse!

**Pull request merged**